pkg/smokescreen/acl/v1/acl.go

```diff
@@ -0,0 +1,230 @@
+package acl
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+)
+
+type Decider interface {
+	Decide(service, host string) (Decision, error)
+}
+
+type ACL struct {
+	Rules            map[string]Rule
+	DefaultRule      *Rule
+	GlobalDenyList   []string
+	GlobalAllowList  []string
+	DisabledPolicies []EnforcementPolicy
+	*logrus.Logger
+}
+
+type Rule struct {
+	Project     string
+	Policy      EnforcementPolicy
+	DomainGlobs []string
+}
+
+type Decision struct {
+	Reason  string
+	Default bool
+	Result  DecisionResult
+	Project string
+}
+
+func New(logger *logrus.Logger, loader Loader, disabledActions []string) (*ACL, error) {
+	acl, err := loader.Load()
+	if err != nil {
+		return nil, err
+	}
+
+	acl.DisablePolicies(disabledActions)
+
+	err = acl.Validate()
+	if err != nil {
+		return nil, err
+	}
+
+	acl.Logger = logger
+
+	if acl.DefaultRule == nil {
+		acl.Warn("no default rule set")
+	}
+	return acl, nil
+}
+
+// Add associates a rule with the specified service after verifying the rule's
+// policy and domains are valid.
+func (acl *ACL) Add(svc string, r Rule) error {
+	err := acl.PolicyDisabled(svc, r.Policy)
+	if err != nil {
+		return err
+	}
+
+	err = acl.ValidateDomains(r.DomainGlobs)
+	if err != nil {
+		return err
+	}
+
+	acl.Rules[svc] = r
+	return nil
+}
+
+// Decide takes uses the rule configured for the given service to determine if
+//   1. The host is in the rule's allowed domain
+//   2. The host has been globally denied
+//   3. The host has been globally allowed
+//   4. There is a default rule for the ACL
+func (acl *ACL) Decide(service, host string) (Decision, error) {
+	var d Decision
+
+	rule := acl.Rule(service)
+	if rule == nil {
+		d.Result = Deny
+		d.Reason = "no rule matched"
+		return d, nil
+	}
+
+	d.Project = rule.Project
+	d.Default = rule == acl.DefaultRule
+
+	// if the host matches any of the rule's allowed domains, allow
+	for _, dg := range rule.DomainGlobs {
+		if hostMatchesGlob(host, dg) {
+			d.Result, d.Reason = Allow, "host matched allowed domain in rule"
+			return d, nil
+		}
+	}
+
+	// if the host matches any of the global deny list, deny
```

Comment:
Just confirming: We want individual services' ACLs to override the global deny list?

Reply:
For better or worse, this is how the implementation works today https://github.com/stripe/smokescreen/blob/master/pkg/smokescreen/acl_loader_v1.go#L38. Given that this is already a rather large PR I was punting on changing the expected behavior. I do agree this is something we should revisit because I'm not sure this is the behavior we actually want.

Reply:
Got it, thanks for taking the time to explain it to me!